### ISSUE TITLE
`SSL_CTX_set_cert_verify_callback(3)` semantic change

### ISSUE THREAD

**Issue by t-j-h** [openssl/openssl#17568]
IMNSHO the mechanism chosen to signal this is rather an unfortunate incompatibility with OpenSSL 1.x.  Previously negative return values from the application verificcation callback (i.e. set via `SSL_CTX_set_cert_verify_callback(3)`) had no special semantics, but were typically indications of an internal error performing the verification, rather than a *normal* verification failure.

A typical application callback would manipulate the X509_STORE_CTX object and then just tail-call the standard `X509_verify_cert(3)`, whose return value was expected to be compatible for use as application callback.

Now there appears to a special case where -1 is special when returned by an application callback, but not by the built-in verification code.  This is pretty bad, and breaks existing code.

I think this design decision needs to be reconsidered.

_Originally posted by @vdukhovni in https://github.com/openssl/openssl/issues/13906#issuecomment-1019343986_

**Comment by vdukhovni** [comment]
My suggestion is to change the signal to be a special value of the X509_STORE_CTX `error` value, rather than overload the return value in kludgey and backwards-incompatible way.

Another option is to mutate the `SSL *` handle to set an indication that verification should be retried.  The `SSL *` can be passed to the application callback as its application context.  This would move such signalling entirely out of the X509 layer, and the SSL retry signalling can be then be based on inspection of the SSL handle state.  Other approaches may be possible, really anything but what we have now in 3.0.

**Comment by vdukhovni** [comment]
Note this should obviate the work-around in 758754966791c537ea95241438454aa86f91f256

**Comment by DDvO** [comment]
In a sense this issue is one of the many instances of #17494.

**Comment by DDvO** [comment]
While I agree that there is some trouble here to be solved somehow,
I do not agree on the initial analysis because the documentation of `SSL_CTX_set_cert_verify_callback.pod` since its introduction in commit c4068186acec0f5cd370f7b5772b41f215903234 of August 2001 clearly says:
> _callback_ should return 1 to indicate verification success and 0 to
indicate verification failure.

So the bug/error is that some callback implementations did not follow the requirements given in the documentation.

**Comment by vdukhovni** [comment]
> So the core bug/error is that some callback implementation(s), at least the one fixed by commit [7587549](https://github.com/openssl/openssl/commit/758754966791c537ea95241438454aa86f91f256), did not adhere to the requirements given in the documentation.

This may be so, but the fact that the built-in default "callback" (i.e. X509_verify_cert()) is the prime example of how to implement the callback, and was typically used as a tail call in custom callbacks, suggests a documentation oversight more than than an intended contract.

This is especially so given that for a long time OpenSSL documentation was *RTFS*, and the FS clearly showed that the calling code treated `0` and `< 0` as equivalent failure conditions.  Indeed it still does for the built-in default verification code path, and the new treatment of negative returns for applications that never expected it to mean `retry` is very unfortunate (infinite loop on internal resource errors, malformed inputs, ...)

This introduces DoS vectors into previously working application code.  If there is to be a new retry signalling mechanism, it needs to be a *new* signal and not a foot-gun for existing applications.

**Comment by DDvO** [comment]
Oh, this indicates an instance of a much more fundamental issue:
outdated/incomplete and in this case even misleading documentation :frowning_face:

**Comment by vdukhovni** [comment]
We can get sidetracked into various larger questions, but I do think that the immediate issue needs attention.  The change in the callback semantics in 1.1.1 -> 3.0 rather needs to be reverted I think.

To signal a desire to retry what's needed is a new mechanism (e.g. custom error code, or perhaps even better a special call on the SSL handle that the application callback can make).

**Comment by t8m** [comment]
> A simpler alternative to the solution options mentioned above would be to use instead of `-1` some value that `X509_verify_cert()` is guaranteed to never return, e.g., `2` or some smaller negative value.
> Yet there may be reasons speaking against this very primitive pragmatic solution option.

That would be absolutely horrendous kludge and I do not think it would go through the OTC. IMO the suggested way of using additional X509_STORE_CTX error value would be the least problematic. But I might not see all possible consequences from that.

**OTC decision needed: Should we revert the semantic change for the cert verify callback return value (on master and 3.0) and if so, what should we replace it with?**

**Comment by DDvO** [comment]
Yes, and the misleading documentation of `SSL_CTX_set_cert_verify_callback()` needs to be corrected as well.

**Comment by rousskov** [comment]
We have a work-in-progress project that is supposed to rely on #13906 OpenSSL API. We have not submitted our changes for the official Squid Project inclusion yet and will use whatever fixed/new/improved API is adopted by the OpenSSL Project if that adoption happens soon.

FWIW, we would very much prefer not to support _two_ APIs: 13906 API already in OpenSSL v3 and whatever new API this issue results in. However, I do not know how flexible OpenSSL Project is with regard to substituting an already published API with a new one. Will that change have to wait until OpenSSL v3.1? v4.0?? If possible, please resolve this quickly so that we can move forward with a _single_ OpenSSL API version we can rely on long-term! Thank you.

**Comment by vdukhovni** [comment]
I agree there's a need to resolve this quickly, and likely by breaking the initial 3.0 API restoring 1.1.1 semantics for the return value, and introducing a new signalling channel, either via the error code, or via a new SSL handle method that mutates a field that is then checked by the state machine.

My preference is actually for the latter, as a matter of layer separation and hygiene, but the former at least avoids surprises for applications that never expected negative return values to retry verification.

**Comment by vdukhovni** [comment]
There's an OTC call tomorrow, perhaps I should join, if that's appropriate...

**Comment by mattcaswell** [comment]
>There's an OTC call tomorrow, perhaps I should join, if that's appropriate...

Unfortunately we ran out of time this week - hopefully next week.

**Comment by t8m** [comment]
> My preference is actually for the latter, as a matter of layer separation and hygiene, but the former at least avoids surprises for applications that never expected negative return values to retry verification.

Wouldn't the verification retry error make a sense outside of the libssl context as well? IMO that would be a reason to introduce such error code into X509_V_ERR namespace. Otherwise yeah, SSL handle method would be better choice.

**Comment by vdukhovni** [comment]
> Wouldn't the verification retry error make a sense outside of the libssl context as well? IMO that would be a reason to introduce such error code into X509_V_ERR namespace. Otherwise yeah, SSL handle method would be better choice.

We're talking about an SSL callback interface, with the SSL state machine driving certificate verification.  So there isn't a more general context in which one would tell the upper layer to retry, or if there were that upper layer's callback API can provide similar facilities.

**Comment by vdukhovni** [comment]
Since the OTC call today did not get to this issue, perhaps we can make some more progress in the discussion prior to the next call.

* Are there substantial objections to revising the API to restore 1.1.1 compatible callback behaviour (-ve means failure, not retry).
* Which alternative "please retry" signalling method would you suggest?

**Comment by DDvO** [comment]
> > Wouldn't the verification retry error make a sense outside of the libssl context as well? IMO that would be a reason to introduce such error code into X509_V_ERR namespace. Otherwise yeah, SSL handle method would be better choice.
> 
> We're talking about an SSL callback interface, with the SSL state machine driving certificate verification. So there isn't a more general context in which one would tell the upper layer to retry, or if there were that upper layer's callback API can provide similar facilities.

Although the current use and the immediate pain in the SSL/TLS implementation,
I agree with @t8m that the concept of possibly retrying a cert validation attempt (with possibly enhanced input or the like) should be applicable more generally.
So I like the idea of solving the issue generally at the X509 verify level, for instance by extending `X509_V_ERR`.

**Comment by t8m** [comment]
The problem with adding a new SSL call is that that requires passing the SSL handle to the callback as the user data because the only other parameter of the callback is the store context. But the SSL_CTX_set_cert_verify_callback is per SSL context and not per SSL handle - how do you alter the arg parameter for the callback to be the SSL handle? That looks fairly impossible to me. The alternative then requires the application when implementing the callback to be able to somehow get to the SSL handle that invoked the callback, which might fairly complicate the callback implementation.

**Comment by vdukhovni** [comment]
The SSL handle is available inside the callback via `SSL_get_ex_data_X509_STORE_CTX_idx(3)`.

**Comment by vdukhovni** [comment]
> Ah, OK then I lean towards adding the SSL call. We might do also something further in master but I'd rather not extend the X509_V_ERR in 3.0.

It would be best if the API in 3.0 (perhaps ultimately LTS) were in fact the long-term form of the preferred way to signal **retry**.

**Comment by t8m** [comment]
**OTC decision: The 13906 should be reverted and alternative mechanism should be provided**

**Comment by vdukhovni** [comment]
Thanks for the OTC decision.  Is someone already working on a patch, or is likely to land in my court?

**Comment by mattcaswell** [comment]
The fellows aren't working on this at the moment. @DDvO are you?

**Comment by DDvO** [comment]
I am not working on this,
and I'd say I should not because, as it turned out, I have too little familiarity with the SSL implementation.

**Comment by hlandau** [comment]
Summarizing the above, the options for the alternative mechanism are:

- A new return code.
- Adding a new `X509_V_*` error code.
- A new SSL API call.

It does feel like a new SSL API call would be the cleanest way to signal this. A possible design would be:
```
int X509_retry(X509_STORE_CTX *);
```

The obvious usage would be something like:
```
{
  if (can_verify_now) {
    return X509_verify_cert(ctx);
  } else {
    X509_retry(ctx);
    return -1;
  }
}
```

But I'm wondering if we should ask people to return the return value of the function instead. This allows us to sidestep the issue of what return value should be used in conjunction with the call, and makes it less likely that the call will be made but the correct return value not used, or vice versa:

```
{
  if (can_verify_now) {
    return X509_verify_cert(ctx);
  } else {
    return X509_retry(ctx);
  }
}
```

`X509_retry` would be documented as returning an unspecified value which should be returned opaquely by the callback.

## Implementation

Actual implementation would probably look something like this:

- `X509_retry` would actually return -1. (If the latter example above is used, there would also be the option of having a new but undocumented value not used by X509_verify_cert)
- The question of what to do if `X509_retry` is used but the wrong return code is used would be resolved as follows:
  - If `X509_retry` is not used, any return value behaves as it does previously, thus backwards compatibility is not affected.
  - If `X509_retry` is used but the return value is not -1 or whatever it returns, validation
    fails as though -1 was returned without retry being requested. This is considered an implementation error and thus conservatively the process fails. We are conservative here because this is a security sensitive callback and misinterpretation of decisions on the part of the application has the potential to be catastrophic.

This is a strawman proposal, feedback is welcomed and desired.

**Comment by t8m** [comment]
@hlandau your proposal is doing the first option in your list which isn't something we want.

The new SSL API call option is implemented in #17825.
